### ISSUE TITLE
_Float16 on amd64

### ISSUE THREAD

**Issue by UltraCoderRU** [ROCmSoftwarePlatform/rocBLAS#1146]
Hello. I've tried to build PyTorch with ROCm support using Clang 9/10 and got error:
```
/opt/rocm-3.3.0/include/rocblas-types.h:55:9: error: _Float16 is not supported on this target
typedef _Float16 rocblas_half;
        ^
```
Then I switched to GCC 8.4 and PyTorch was build successfully.

Then I created minimal project using rocBLAS:
```
#include <rocblas-types.h>
int main() {
	rocblas_half half;
	return 0;
}
```
and got the same error. Again, it compiles with GCC.

According to Clang [documentation](https://clang.llvm.org/docs/LanguageExtensions.html#half-precision-floating-point), it seems, that `_Float16` is not yet supported on amd64, and so this condition is incorrect:
```
// Clang supports _Float16 on C11 and C++11
// GCC does not currently support _Float16 on amd64
/*! \brief Represents a 16 bit floating point number. */
#if __clang_major__ >= 9 && (__STDC_VERSION__ >= 201112L || __cplusplus >= 201103L)
```
Sorry, if I missed something, never worked with ROCm before.

### Environment
| Hardware | description |
|-----|-----|
| GPU | AMD Radeon RX Vega 56 |
| CPU | AMD Ryzen 7 1700 |

| Software | version |
|-----|-----|
| ROCm | v3.3.0 (from official repository)|

**Comment by leekillough** [comment]
Clang 9+ added support for `_Float16`:

https://reviews.llvm.org/D33719

Since we could not determine precisely which minor version of Clang 9 had this change, the current logic uses Clang 10 as the cutoff:
```
#if __clang_major__ >= 10 && (__STDC_VERSION__ >= 201112L || __cplusplus >= 201103L)
```

**Comment by UltraCoderRU** [comment]
Changes you are referring to are from 2017, but later in 2019 `_Float16` was disabled for non ARM/SPIR Targets ([commit](https://github.com/llvm/llvm-project/commit/1d1d438e8e721d8cef58b638fc304c2ff87094f0)).

Even in [10.x branch](https://github.com/llvm/llvm-project/blob/release/10.x/clang/lib/Basic/TargetInfo.cpp#L38) `TargetInfo::HasFloat16==false` and this flag doesn't reset in
[X86TargetInfo](https://github.com/llvm/llvm-project/blob/release/10.x/clang/lib/Basic/Targets/X86.h#L42). So `_Float16` support is still limited to few architectures.

Again, from [documentation](https://clang.llvm.org/docs/LanguageExtensions.html#half-precision-floating-point):
> _Float16 will be supported on more targets as they define ABIs for it.

**Comment by leekillough** [comment]
Will be fixed in 3.5: [PR# 1148](https://github.com/ROCmSoftwarePlatform/rocBLAS/pull/1148)
